contracts/v2/marketplace/TokenMarketplaceV2.sol

```diff
@@ -0,0 +1,509 @@
+
+    Offer storage offer = offers[_tokenId];
+
+    uint256 winningOffer = offer.offer;
+    // Check valid offer and offer not replaced whilst inflight
+    require(winningOffer > 0 && winningOffer >= _acceptedAmount, "Offer amount not satisfied");
+
+    address winningBidder = offers[_tokenId].bidder;
+
+    delete offers[_tokenId];
+
+    // Get edition no.
+    uint256 editionNumber = kodaAddress.editionOfTokenId(_tokenId);
+
+    _handleFunds(editionNumber, winningOffer, currentOwner);
+
+    kodaAddress.safeTransferFrom(msg.sender, winningBidder, _tokenId);
+
+    emit BidAccepted(_tokenId, currentOwner, winningBidder, winningOffer);
+  }
+
+  function _refundHighestBidder(uint256 _tokenId) internal {
+    // Get current highest bidder
+    address currentHighestBidder = offers[_tokenId].bidder;
+
+    // Get current highest bid amount
+    uint256 currentHighestBiddersAmount = offers[_tokenId].offer;
+
+    if (currentHighestBidder != address(0) && currentHighestBiddersAmount > 0) {
+
+      // Clear out highest bidder
+      delete offers[_tokenId];
+
+      // Refund it
+      currentHighestBidder.transfer(currentHighestBiddersAmount);
+    }
+  }
+
+  //////////////////////////
+  // User Listing Actions //
+  //////////////////////////
+
+  function listToken(uint256 _tokenId, uint256 _listingPrice)
+  public
+  whenNotPaused {
+    // Check ownership before listing
+    address tokenOwner = kodaAddress.ownerOf(_tokenId);
+    require(tokenOwner == msg.sender, "Not token owner");
+
+    // Check price over min bid
+    require(_listingPrice >= minBidAmount, "Listing price not enough");
+
+    // List the token
+    listings[_tokenId] = Listing({
+    price : _listingPrice,
+    seller : msg.sender
+    });
+
+    emit TokenListed(_tokenId, msg.sender, _listingPrice);
+  }
+
+  function delistToken(uint256 _tokenId)
+  public
+  whenNotPaused {
+
+    // check listing found
+    require(listings[_tokenId].seller != address(0), "No listing found");
+
+    // check owner is msg.sender
+    require(kodaAddress.ownerOf(_tokenId) == msg.sender, "Only the current owner can delist");
+
+    _delistToken(_tokenId);
+  }
+
+  function buyToken(uint256 _tokenId)
+  public
+  payable
+  nonReentrant
+  whenNotPaused {
+    Listing memory listing = listings[_tokenId];
+
+    // check token is listed
+    require(listing.seller != address(0), "No listing found");
+
+    // check current owner is the lister as it may have changed hands
+    address currentOwner = kodaAddress.ownerOf(_tokenId);
+    require(listing.seller == currentOwner, "Listing not valid, token owner has changed");
+
+    // check listing satisfied
+    uint256 listingPrice = listing.price;
+    require(msg.value == listingPrice, "List price not satisfied");
+
+    // Get edition no.
+    uint256 editionNumber = kodaAddress.editionOfTokenId(_tokenId);
+
+    // refund any open offers on it
+    Offer memory offer = offers[_tokenId];
+    if (offer.bidder != address(0)) {
```

Comment:
unnecessary check as it's already done in `_refundHighestBidder `

Reply:
Fixed